akd/src/storage/memory.rs

```diff
@@ -266,6 +266,43 @@ impl Storage for AsyncInMemoryDatabase {
         }
         Ok(map)
     }
+
+    async fn get_epoch_lte_epoch(
+        &self,
+        node_label: crate::node_state::NodeLabel,
+        epoch_in_question: u64,
+    ) -> Result<u64, StorageError> {
```

Comment:
So I guess the cool thing about doing this "linear search" for the correct epoch is that it can be done efficiently with a MySQL query, but may be less efficient if we are just relying on a simple set/get interface. Would it be better to find a middle-ground here with some pointer-based solution? Or perhaps we are not too worried about the performance impact of this function because it is not going to be called in lookup or publish?

Reply:
Yeah so the key history and audit pero queries do make this call, but we could probably figure out a nifty way to cache it but I think this is a good starting point at least.

I agree that it's not perfect but it completely mitigates the unbounded storage problems.

Reply:
Agreed, good point. Let's merge this for now then, and if we run into issues with performance on key history / audit, we can always revisit.

Thanks!!

Reply:
I believe most key-value storage systems provide some sort of key scan or match API that would make regex matching or prefix searching possible. 

It would then be up to the storage layer implementer to use an encoding scheme that would allow for the most efficient searching for the specific database :)